File: app/src/api/deconstruct.ts
import { USE_MOCK, requestResponse } from './http'

export interface DeconstructionBeat {
  position: number
  label: string
  excerpt: string
}

export interface DeconstructionChapter {
  index: number
  title: string
  word_count: number
  paragraph_count: number
  dialogue_ratio: number
  summary: string
  beats: DeconstructionBeat[]
  confidence: string
}

export interface DeconstructionResult {
  parser: string
  stats: {
    total_words: number
    chapter_count: number
    average_chapter_words: number
    median_chapter_words: number
  }
  chapters: DeconstructionChapter[]
  rhythm_nodes: Array<{ chapter_index: number; position: number; label: string; intensity: number }>
  payoff_distribution: Array<{
    range: string
    chapter_from: number | null
    chapter_to: number | null
    score: number
    peak_chapters: number[]
  }>
  warnings: string[]
  disclaimer: string
}

function mockDeconstruction(text: string): DeconstructionResult {
  const normalized = text.replace(/\r\n?/g, '\n').trim()
  const lines = normalized.split('\n')
  const heading = /^\s*(第[零一二三四五六七八九十百千万\d]+[章节回部卷].{0,80}|Chapter\s+\d+.*|序章|楔子|尾声|番外.*)\s*$/i
  const starts = lines.flatMap((line, index) => {
    const title = line.replace(/^\s{0,3}#{1,6}\s+/, '').trim().replace(/#+$/, '').trim()
    return heading.test(title) ? [{ index, title }] : []
  })
  const ranges = starts.length ? starts.map((start, index) => ({ title: start.title, body: lines.slice(start.index + 1, starts[index + 1]?.index ?? lines.length).join('\n') })) : [{ title: '全文', body: normalized }]
  const count = (value: string) => Array.from(value).filter((char) => /[A-Za-z0-9\u3400-\u9fff]/.test(char)).length
  const chapters = ranges.map((range, index) => {
    const paragraphs = range.body.split(/\n\s*\n/).map((part) => part.trim()).filter(Boolean)
    const body = paragraphs.join(' ')
    const dialogueWords = Array.from(body.matchAll(/[「“"]([^」”"\n]+)[」”"]/g)).reduce((sum, match) => sum + count(match[1] ?? ''), 0)
    const words = count(body)
    return {
      index: index + 1,
      title: range.title,
      word_count: words,
      paragraph_count: paragraphs.length,
      dialogue_ratio: Number(Math.min(1, dialogueWords / Math.max(1, words)).toFixed(3)),
      summary: body.slice(0, 180),
      beats: paragraphs.length ? [{ position: 0.5, label: '场景推进', excerpt: paragraphs[0].slice(0, 120) }] : [],
      confidence: range.title === '全文' ? 'low' : 'high'
    }
  })
  const total = count(normalized)
  return {
    parser: 'MOCK',
    stats: { total_words: total, chapter_count: chapters.length, average_chapter_words: Math.round(total / Math.max(1, chapters.length)), median_chapter_words: chapters[Math.floor(chapters.length / 2)]?.word_count ?? 0 },
    chapters,
    rhythm_nodes: chapters.map((chapter) => ({ chapter_index: chapter.index, position: 0.5, label: '场景推进', intensity: 20 })),
    payoff_distribution: Array.from({ length: 10 }, (_, index) => ({ range: `${index * 10 + 1}-${(index + 1) * 10}%`, chapter_from: null, chapter_to: null, score: 0, peak_chapters: [] })),
    warnings: ['当前为演示模式，结果不会上传服务器。', '节奏与爽点信号仅供参考，请结合原文人工复核。'],
    disclaimer: '仅供结构学习参考，不构成版权许可、审查或抄袭建议。'
  }
}

export const deconstructionApi = {
  async analyze(file: File): Promise<DeconstructionResult> {
    if (USE_MOCK) return mockDeconstruction(await file.text())
    const query = new URLSearchParams({ filename: file.name })
    const response = await requestResponse(`/analysis/deconstruct?${query}`, {
      method: 'POST', body: file, headers: { 'Content-Type': 'application/octet-stream' }
    })
    return response.json() as Promise<DeconstructionResult>
  }
}

File: server/api/deconstruct.py
"""Reference-book deconstruction without persistence or model calls."""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from pydantic import BaseModel, ConfigDict

from api.auth import get_current_user
from db.models_core import User
from services.book_deconstruction import MAX_UPLOAD_BYTES, DeconstructionError, analyze_book

router = APIRouter()


class DeconstructionStats(BaseModel):
    total_words: int
    chapter_count: int
    average_chapter_words: int
    median_chapter_words: int


class DeconstructionBeat(BaseModel):
    position: float
    label: str
    excerpt: str


class DeconstructionChapter(BaseModel):
    index: int
    title: str
    word_count: int
    paragraph_count: int
    dialogue_ratio: float
    summary: str
    beats: list[DeconstructionBeat]
    confidence: str


class RhythmNode(BaseModel):
    chapter_index: int
    position: float
    label: str
    intensity: int


class PayoffBucket(BaseModel):
    range: str
    chapter_from: int | None
    chapter_to: int | None
    score: int
    peak_chapters: list[int]


class DeconstructionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    parser: str
    stats: DeconstructionStats
    chapters: list[DeconstructionChapter]
    rhythm_nodes: list[RhythmNode]
    payoff_distribution: list[PayoffBucket]
    warnings: list[str]
    disclaimer: str


@router.post("/deconstruct", response_model=DeconstructionResponse)
async def deconstruct_reference_book(
    request: Request,
    filename: str = Query(min_length=1, max_length=255),
    _user: User = Depends(get_current_user),
) -> DeconstructionResponse:
    """Analyze a reference manuscript in memory; never persist or enqueue its text."""
    content_length = request.headers.get("content-length")
    if content_length and content_length.isdigit() and int(content_length) > MAX_UPLOAD_BYTES:
        raise HTTPException(status_code=422, detail={"code": "FILE_TOO_LARGE", "reason": "单个文件不能超过 10 MB。"})
    raw = bytearray()
    async for chunk in request.stream():
        if len(raw) + len(chunk) > MAX_UPLOAD_BYTES:
            raise HTTPException(status_code=422, detail={"code": "FILE_TOO_LARGE", "reason": "单个文件不能超过 10 MB。"})
        raw.extend(chunk)
    try:
        result = analyze_book(bytes(raw), filename)
    except DeconstructionError as exc:
        raise HTTPException(status_code=422, detail={"code": exc.code, "reason": str(exc)}) from exc
    except Exception as exc:  # Keep malformed archives from becoming an opaque 500.
        raise HTTPException(status_code=422, detail={"code": "PARSE_ERROR", "reason": "文件无法安全解析。"}) from exc
    return DeconstructionResponse.model_validate(result)


__all__ = ["router"]
